### PR TITLE
planner: fix the problem of using `enum like 'x%'` to build the wrong range

### LINKED ISSUE #27130: The result of query using the tabledual plan is not correct
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t( a enum('y','b','Abc','null'),b enum('y','b','Abc','null'),key(a));
insert into t values("Abc","Abc");
select * from t where a like "A%";  -- expect got Abc, Abc; but got Empty set
/*
+-------------+---------+------+---------------+---------------+
| id          | estRows | task | access object | operator info |
+-------------+---------+------+---------------+---------------+
| TableDual_5 | 0.00    | root |               | rows:0        |
+-------------+---------+------+---------------+---------------+
*/
select * from t where b like "A%"; -- correctly
/*
+-------------------------+---------+-----------+---------------+--------------------------------+
| id                      | estRows | task      | access object | operator info                  |
+-------------------------+---------+-----------+---------------+--------------------------------+
| TableReader_7           | 0.00    | root      |               | data:Selection_6               |
| └─Selection_6           | 0.00    | cop[tikv] |               | like(rs.t.b, "A%", 92)         |
|   └─TableFullScan_5     | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+---------+-----------+---------------+--------------------------------+
*/

```
### 2. What did you expect to see? (Required)
tabledual plan result is correct
### 3. What did you see instead (Required)
tabledual plan result is not correct
### 4. What is your TiDB version? (Required)

Release Version: v5.2.0-alpha-549-g6720fcd40
Edition: Community
Git Commit Hash: 6720fcd40ff90f11c2f678bb42bb288474854c19
Git Branch: master
UTC Build Time: 2021-08-09 03:22:37
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve?

Issue Number: close #27130 

Problem Summary:

We used `enum like 'x%'` to build the wrong range. In fact, we do not support the use of `enum like 'x%'` as access condition of IndexReader.

### What is changed and how it works?

Let `enum like 'x%'` be filter condition of IndexReader.

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
planner: fix the problem of using `enum like 'x%'` to build the wrong range
```

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: util/ranger/checker.go [hunk 1]
```diff
@@ -145,12 +146,22 @@ func (c *conditionChecker) checkLikeFunc(scalar *expression.ScalarFunction) bool
...
+			// see https://github.com/pingcap/tidb/issues/27130 for more details
+			if scalar.GetArgs()[0].GetType().Tp == mysql.TypeEnum {
+				return false
+			}
 			if i != len(patternStr)-1 {
 				c.shouldReserve = true
 			}
 			break
 		}
 		if patternStr[i] == '_' {
```
Comment (Reminiscent) [inline util/ranger/checker.go:159]: Please add some tests for this case. Rest LGTM

**Review by Reminiscent — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**